File: ast/lex.go
package ast

import (
	"strconv"
	"strings"
)

// Lex converts input into a token stream terminated by an EOF token, plus any
// lexical diagnostics (AGENTS.md §3).
//
// Lexing recovers rather than stopping: a malformed string literal still yields
// a STRING token so the parser can carry on and report the problems further
// along the line, and an unrecognized character is reported and skipped. Nothing
// unrecognized is ever carried into the token stream — invariant 2 starts here.
func Lex(input string) ([]Token, []Diagnostic) {
	l := &lexer{src: []rune(input)}
	l.run()
	return l.toks, l.diags
}

type lexer struct {
	src   []rune
	pos   int
	toks  []Token
	diags []Diagnostic
}

func (l *lexer) run() {
	for l.pos < len(l.src) {
		c := l.src[l.pos]
		switch {
		case isSpace(c):
			l.pos++
		case c == '(':
			l.emit(LPAREN, "(", l.pos, l.pos+1)
			l.pos++
		case c == ')':
			l.emit(RPAREN, ")", l.pos, l.pos+1)
			l.pos++
		case c == '"':
			l.lexString()
		case isIdentStart(c):
			l.lexIdent()
		case isDigit(c) || (c == '-' && l.pos+1 < len(l.src) && isDigit(l.src[l.pos+1])):
			l.lexNumber()
		default:
			if n := l.operatorLen(); n > 0 {
				l.emit(OP, string(l.src[l.pos:l.pos+n]), l.pos, l.pos+n)
				l.pos += n
				continue
			}
			l.diag(CodeUnexpectedToken, Span{l.pos, l.pos + 1},
				"unexpected character "+strconv.QuoteRune(c))
			l.pos++
		}
	}
	l.emit(EOF, "", len(l.src), len(l.src))
}

// operatorLen returns the length of the operator at the cursor, matching
// longest-first so that != , !~ and <= are never split (AGENTS.md §3.1).
func (l *lexer) operatorLen() int {
	rest := l.src[l.pos:]
	if len(rest) >= 2 {
		switch string(rest[:2]) {
		case "!=", "!~", "<=", ">=":
			return 2
		}
	}
	switch rest[0] {
	case '=', '~', '<', '>':
		return 1
	}
	return 0
}

func (l *lexer) lexIdent() {
	start := l.pos
	l.pos++
	for l.pos < len(l.src) && isIdentPart(l.src[l.pos]) {
		l.pos++
	}
	text := string(l.src[start:l.pos])
	kind := IDENT
	switch strings.ToLower(text) {
	case "and":
		kind = AND
	case "or":
		kind = OR
	case "not":
		kind = NOT
	case "true":
		kind = TRUE
	case "false":
		kind = FALSE
	}
	l.emit(kind, text, start, l.pos)
}

func (l *lexer) lexNumber() {
	start := l.pos
	if l.src[l.pos] == '-' {
		l.pos++
	}
	for l.pos < len(l.src) && isDigit(l.src[l.pos]) {
		l.pos++
	}
	if l.pos+1 < len(l.src) && l.src[l.pos] == '.' && isDigit(l.src[l.pos+1]) {
		l.pos++
		for l.pos < len(l.src) && isDigit(l.src[l.pos]) {
			l.pos++
		}
	}
	text := string(l.src[start:l.pos])
	n, err := strconv.ParseFloat(text, 64)
	if err != nil {
		// Unreachable for the pattern above, but a malformed number must not
		// become a passthrough token.
		l.diag(CodeUnexpectedToken, Span{start, l.pos}, "invalid number "+text)
		return
	}
	tok := Token{Kind: NUMBER, Text: text, Num: n, Span: Span{start, l.pos}}
	l.toks = append(l.toks, tok)
}

func (l *lexer) lexString() {
	start := l.pos
	l.pos++ // opening quote
	var sb strings.Builder
	for {
		if l.pos >= len(l.src) {
			l.diag(CodeUnterminatedString, Span{start, len(l.src)},
				"string literal is not closed")
			l.emitString(sb.String(), start, len(l.src))
			return
		}
		c := l.src[l.pos]
		if c == '"' {
			l.pos++
			l.emitString(sb.String(), start, l.pos)
			return
		}
		if c != '\\' {
			sb.WriteRune(c)
			l.pos++
			continue
		}
		if l.pos+1 >= len(l.src) {
			l.diag(CodeUnterminatedString, Span{start, len(l.src)},
				"string literal is not closed")
			l.emitString(sb.String(), start, len(l.src))
			l.pos = len(l.src)
			return
		}
		switch e := l.src[l.pos+1]; e {
		case '"', '\\':
			sb.WriteRune(e)
		case 'n':
			sb.WriteRune('\n')
		case 't':
			sb.WriteRune('\t')
		case 'r':
			sb.WriteRune('\r')
		default:
			l.diag(CodeInvalidEscape, Span{l.pos, l.pos + 2},
				"unknown escape \\"+string(e))
			sb.WriteRune(e) // recovery: keep the character, keep parsing
		}
		l.pos += 2
	}
}

func (l *lexer) emitString(value string, start, end int) {
	l.toks = append(l.toks, Token{
		Kind: STRING,
		Text: string(l.src[start:end]),
		Str:  value,
		Span: Span{start, end},
	})
}

func (l *lexer) emit(kind TokenKind, text string, start, end int) {
	l.toks = append(l.toks, Token{Kind: kind, Text: text, Span: Span{start, end}})
}

func (l *lexer) diag(code string, span Span, msg string) {
	l.diags = append(l.diags, Diagnostic{Code: code, Message: msg, Span: span})
}

func isSpace(c rune) bool { return c == ' ' || c == '\t' || c == '\r' || c == '\n' }
func isDigit(c rune) bool { return c >= '0' && c <= '9' }

func isIdentStart(c rune) bool {
	return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

func isIdentPart(c rune) bool { return isIdentStart(c) || isDigit(c) || c == '.' }

File: js/packages/core/src/lex.js
import { codes, diagnostic } from './diagnostic.js'

/**
 * Token kinds. The spellings are normative: the conformance protocol reports
 * them verbatim (AGENTS.md §3.1, §11).
 */
export const kinds = {
  IDENT: 'IDENT',
  STRING: 'STRING',
  NUMBER: 'NUMBER',
  TRUE: 'TRUE',
  FALSE: 'FALSE',
  AND: 'AND',
  OR: 'OR',
  NOT: 'NOT',
  OP: 'OP',
  LPAREN: 'LPAREN',
  RPAREN: 'RPAREN',
  EOF: 'EOF',
}

const keywords = {
  and: kinds.AND,
  or: kinds.OR,
  not: kinds.NOT,
  true: kinds.TRUE,
  false: kinds.FALSE,
}

const isSpace = (c) => c === ' ' || c === '\t' || c === '\r' || c === '\n'
const isDigit = (c) => c >= '0' && c <= '9'
const isIdentStart = (c) => c === '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
const isIdentPart = (c) => isIdentStart(c) || isDigit(c) || c === '.'

/** True when the token can stand in value position (AGENTS.md §5). */
export function isValueToken(token) {
  return (
    token.kind === kinds.STRING ||
    token.kind === kinds.NUMBER ||
    token.kind === kinds.TRUE ||
    token.kind === kinds.FALSE
  )
}

/**
 * Lex converts input into a token stream terminated by an EOF token, plus any
 * lexical diagnostics (AGENTS.md §3).
 *
 * Positions are codepoint offsets, which is why this walks an array of
 * codepoints rather than indexing the string: a JavaScript string is UTF-16, and
 * an emoji in a filter value must not shift every span after it.
 *
 * Lexing recovers rather than stopping: a malformed string literal still yields
 * a STRING token so the parser can carry on, and an unrecognized character is
 * reported and skipped. Nothing unrecognized is ever carried into the token
 * stream — invariant 2 starts here.
 */
export function lex(input) {
  const src = Array.from(input)
  const tokens = []
  const diagnostics = []
  let i = 0

  const emit = (kind, text, start, end, extra) =>
    tokens.push({ kind, text, span: [start, end], ...extra })

  while (i < src.length) {
    const c = src[i]

    if (isSpace(c)) {
      i++
    } else if (c === '(') {
      emit(kinds.LPAREN, '(', i, i + 1)
      i++
    } else if (c === ')') {
      emit(kinds.RPAREN, ')', i, i + 1)
      i++
    } else if (c === '"') {
      i = lexString(src, i, tokens, diagnostics)
    } else if (isIdentStart(c)) {
      const start = i
      i++
      while (i < src.length && isIdentPart(src[i])) i++
      const text = src.slice(start, i).join('')
      emit(keywords[text.toLowerCase()] ?? kinds.IDENT, text, start, i)
    } else if (isDigit(c) || (c === '-' && isDigit(src[i + 1] ?? ''))) {
      const start = i
      if (src[i] === '-') i++
      while (i < src.length && isDigit(src[i])) i++
      if (src[i] === '.' && isDigit(src[i + 1] ?? '')) {
        i++
        while (i < src.length && isDigit(src[i])) i++
      }
      const text = src.slice(start, i).join('')
      emit(kinds.NUMBER, text, start, i, { num: Number(text) })
    } else {
      const n = operatorLength(src, i)
      if (n > 0) {
        emit(kinds.OP, src.slice(i, i + n).join(''), i, i + n)
        i += n
      } else {
        diagnostics.push(
          diagnostic(codes.unexpectedToken, [i, i + 1], `unexpected character ${JSON.stringify(c)}`),
        )
        i++
      }
    }
  }

  emit(kinds.EOF, '', src.length, src.length)
  return { tokens, diagnostics }
}

/**
 * Operators are matched longest-first so that !=, !~ and <= are never split
 * (AGENTS.md §3.1). A bare ! is not an operator.
 */
function operatorLength(src, i) {
  const two = src[i] + (src[i + 1] ?? '')
  if (two === '!=' || two === '!~' || two === '<=' || two === '>=') return 2
  const one = src[i]
  if (one === '=' || one === '~' || one === '<' || one === '>') return 1
  return 0
}

function lexString(src, start, tokens, diagnostics) {
  let i = start + 1
  let value = ''

  const finish = (end) => {
    tokens.push({
      kind: kinds.STRING,
      text: src.slice(start, end).join(''),
      str: value,
      span: [start, end],
    })
    return end
  }

  for (;;) {
    if (i >= src.length) {
      diagnostics.push(
        diagnostic(codes.unterminatedString, [start, src.length], 'string literal is not closed'),
      )
      return finish(src.length)
    }
    const c = src[i]
    if (c === '"') return finish(i + 1)
    if (c !== '\\') {
      value += c
      i++
      continue
    }
    if (i + 1 >= src.length) {
      diagnostics.push(
        diagnostic(codes.unterminatedString, [start, src.length], 'string literal is not closed'),
      )
      return finish(src.length)
    }
    const e = src[i + 1]
    switch (e) {
      case '"':
      case '\\':
        value += e
        break
      case 'n':
        value += '\n'
        break
      case 't':
        value += '\t'
        break
      case 'r':
        value += '\r'
        break
      default:
        diagnostics.push(
          diagnostic(codes.invalidEscape, [i, i + 2], `unknown escape \\${e}`),
        )
        value += e // recovery: keep the character, keep parsing
    }
    i += 2
  }
}

/**
 * The token's value as the conformance protocol reports it: a number for
 * NUMBER, a boolean for TRUE/FALSE, the unescaped contents for STRING, and the
 * source text otherwise.
 */
export function tokenValue(token) {
  switch (token.kind) {
    case kinds.NUMBER:
      return token.num
    case kinds.TRUE:
      return true
    case kinds.FALSE:
      return false
    case kinds.STRING:
      return token.str
    default:
      return token.text
  }
}
